sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnInvokers.java

```diff
@@ -310,78 +357,100 @@ protected StackManipulation invokeTargetMethod(MethodDescription instrumentedMet
       parameters.add(MethodVariableAccess.REFERENCE.loadOffset(1));
       // Push the extra arguments in their actual order.
       StackManipulation pushExtraContextFactory = MethodVariableAccess.REFERENCE.loadOffset(2);
-      for (DoFnSignature.ProcessElementMethod.Parameter param : signature.extraParameters()) {
+      for (DoFnSignature.Parameter param : signature.extraParameters()) {
         parameters.add(
             new StackManipulation.Compound(
                 pushExtraContextFactory,
-                MethodInvocation.invoke(EXTRA_CONTEXT_FACTORY_METHODS.get(param))));
+                MethodInvocation.invoke(EXTRA_CONTEXT_FACTORY_METHODS.get(param)),
+                // ExtraContextFactory.restrictionTracker() returns a RestrictionTracker,
+                // but the @ProcessElement method expects a concrete subtype of it.
+                // Insert a downcast.
+                (param == DoFnSignature.Parameter.RESTRICTION_TRACKER)
+                    ? TypeCasting.to(
+                        new TypeDescription.ForLoadedType(signature.trackerT().getRawType()))
+                    : StackManipulation.Trivial.INSTANCE));
       }
 
       return new StackManipulation.Compound(
           // Push the parameters
           new StackManipulation.Compound(parameters),
           // Invoke the target method
           wrapWithUserCodeException(
-              MethodDelegationBinder.MethodInvoker.Simple.INSTANCE.invoke(targetMethod)),
-          // Return from the instrumented method
-          MethodReturn.VOID);
+              new StackManipulation.Compound(
+                  MethodDelegationBinder.MethodInvoker.Simple.INSTANCE.invoke(targetMethod),
+                  // Return from the instrumented @ProcessElement method:
+                  // if it returns void, then return null (meaning don't resume),
+                  // otherwise return the ProcessContinuation it returned.
+                  signature.hasReturnValue()
+                      ? MethodReturn.returning(targetMethod.getReturnType().asErasure())
+                      : new StackManipulation.Compound(
+                          MethodInvocation.invoke(PROCESS_CONTINUATION_STOP_METHOD),
+                          MethodReturn.REFERENCE))));
     }
   }
 
   /**
-   * Implements {@link DoFnInvoker#invokeStartBundle} or {@link DoFnInvoker#invokeFinishBundle} by
-   * delegating respectively to the {@link StartBundle} and {@link FinishBundle} methods.
+   * Delegates to the given method, wrapping the call into a UserCodeException and optionally
+   * reparing parameters.
```

Comment:
"reparing" (I assume repairing) and also not sure exactly what that means. How are parameters damaged?